Add comprehensive filter functionality for changes list

Gr8z opened the pull request:
<!--
What GitHub Desktop issue does this PR address (for example, #1234)?
If you have not created an issue for your PR, please search the issue tracker to see if there is an existing issue that aligns with your PR, or open a new issue for discussion.
-->

## Description
<!--
A summary of the changes made along with any other information that would be helpful to a reviewer such as potential tradeoffs or alternative approaches you considered.
-->
This pull request introduces a new feature to filter changes in the repository by file type and status (new, modified, deleted, unstaged, and staged files). The most important changes include updates to the state management, UI components, and dispatcher logic to support these new filters.

### State Management Updates:
* Added new state properties (`filterNewFiles`, `filterModifiedFiles`, `filterDeletedFiles`, `filterUnstagedFiles`) to `IChangesState` and initialized them in `getInitialRepositoryState`. [[1]](diffhunk://#diff-7aaedfe8e3adadaa69f4dead6a9a93ee2cea03050bf54b80d131162c38412d18R779-R790) [[2]](diffhunk://#diff-e1c08110c1479b61e9fdee22a1df33e750c04c2870fa2fe392058a897aeef6a7R323-R326)
* Updated the `createState` helper in tests to include the new filter properties.

### UI Enhancements:
* Expanded the `FilterChangesList` component to include checkboxes for the new filters, dynamically calculate filter counts, and display the number of active filters. [[1]](diffhunk://#diff-6e4b6c6a393db6cd6ec7ca33302a11e217b008a45c35500719abb92fb2b9bceeR355-R401) [[2]](diffhunk://#diff-6e4b6c6a393db6cd6ec7ca33302a11e217b008a45c35500719abb92fb2b9bceeL1301-R1395) [[3]](diffhunk://#diff-6e4b6c6a393db6cd6ec7ca33302a11e217b008a45c35500719abb92fb2b9bceeR1407-R1426) [[4]](diffhunk://#diff-6e4b6c6a393db6cd6ec7ca33302a11e217b008a45c35500719abb92fb2b9bceeL1333-R1437)
* Adjusted the "no results" message to account for the new filters and their combinations.
* Updated the `ChangesSidebar` to pass the new filter states to the `FilterChangesList` component.

### Dispatcher and Store Logic:
* Added methods in `AppStore` and `Dispatcher` to handle updates to the new filter states (`_setFilterNewFiles`, `_setFilterModifiedFiles`, etc.). [[1]](diffhunk://#diff-b238c18a99c08ab800e422715cc02105067f5f1394027f489e52d7fb15d6ca74R8333-R8369) [[2]](diffhunk://#diff-c8d5cceaa76fd0fe1490c522f46fa1c4b804fffe63ca13acaf0b272578a40b59R4015-R4045)

These changes collectively enhance the user experience by allowing more granular control over the displayed changes in the repository.

### Screenshots
![image](https://github.com/user-attachments/assets/d933c898-b6da-411a-b35a-423e3de984e3)
![image](https://github.com/user-attachments/assets/4e3d3e50-cdb2-4cbd-921c-830877144f85)
![image](https://github.com/user-attachments/assets/58bb4511-4276-4098-a9d4-4a79c12d7fe0)


<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes: Added comprehensive filter functionality for the changes list, allowing users to filter files by status (modified, new, deleted, untracked) for better repository management.

Gr8z committed:
Add comprehensive filter functionality for changes list

- Implement robust filtering for working directory file changes
- Add filter state management to app store and repository cache
- Update UI components (sidebar, filter-changes-list) for filter support
- Enhance dispatcher with filter-related actions
- Add comprehensive test suite covering all filter scenarios
- Include unit tests and integration tests for filter functionality

Gr8z committed:
more filters added to the feature set

Gr8z committed:
add more tests

Gr8z committed:
fix: new files logic

Gr8z committed:
Update filter-changes-list.tsx

Gr8z committed:
remove extra padding from the bottom

Gr8z commented:
Why are GH Actions not running on this PR?

tidy-dev commented:
@Gr8z Thank you for wanting to contribute. 

I will bring this up for team discussion

And, a few things off hand:
- PR's that do not solve an existing issue with the "help-wanted" label are not likely to be merged. It is unclear that the change is a largely desired by the user base - in this case we want to ensure we are adding meaningful filters not a filters for sake of filtering. Secondly, we don't want to see you commit your time to a PR that will not get approved.
- GitHub Desktop does not stage files when you check them. It stages the files upon the user hitting commit. Thus "staged" and "unstaged" is inaccurate terminology; we would need to keep along the lines of "Included in commit"/"Excluded in commit".
- Lastly, just curious why you chose filtering via exclusion versus inclusion? Aka, having all filters checked where unchecking narrows the filter as opposed to starting with filters unchecked and checking a filter option narrows the search. I can see a person saying "I don't want to see deleted files" and thus this being quicker, but equally a user saying "I only want to see deleted files" and thus needing to uncheck multiple. Each direction indirectly gives preference to a particular flow. This feels like something we would want to base on user research on what problems are most common that a filter could solve.

Gr8z commented:
Hi @tidy-dev, thank you for the detailed feedback.

Regarding the issue requirement - You're absolutely right. I should have first opened an issue to discuss the need and gather feedback on whether this feature would be valuable to the user base. I'll create an issue to start that discussion.

Thank you for the clarification about GitHub Desktop's staging behavior. I'll revise the terminology to use "Included in commit"/"Excluded from commit".

**Filter approach**—I did not go for exclusion; I think the screenshot might be misleading. The default state is all unchecked. Filters can be included when you select them, and the condition for multiple filters is "OR." That means if I select Deleted and Modified filters, it will show me Deleted "OR" modified files. 

However, I agree that this assumption should be validated with user research. We should understand:

- What are the most common filtering needs?
- Do users more often want to exclude certain types or focus on specific types?
- What mental model do users have when thinking about filtering their changes?

Would you recommend:

- Pausing this PR while we gather user feedback through an issue discussion?
- Or would you prefer I update the PR with the correct terminology first, so we can better evaluate the implementation while the user research is conducted?

I'm happy to take either approach and appreciate your guidance on making this contribution more aligned with the project's needs.

tidy-dev commented:
> Pausing this PR while we gather user feedback through an issue discussion

Yes. Pausing for now as I do not want to utilize your time on something that will likely take quite some time to garner feedback. Once I discuss it with my team, I will update on if we think we might move forward with it soon or it should be closed until more user feedback is obtained. Thank you for being so understanding. :)

tidy-dev commented:
@Gr8z We discussed this today in a team sync and filtering by file status is one of the iterations that we had considered being a next iteration as it was requested in the [original filtering changes issue](https://github.com/desktop/desktop/issues/6407). Thus, there is prior intention for this filter to be added. We may like to move forward with, but have tabled it for further discussion in an upcoming sync in mid June. 

Thus, I moved it to draft instead of closing it. 

Things we know without in depth review: 
- Previously mentioned "Included in commit"/"Excluded in commit" copy
- A "Clear all" link or button at the bottom of the filter options. 
- The desktop/desktop work for adding in stats tracking on usage of these new filters.  (_Search `incrementMetric` for prior art)_
- Updating the "No results" message to have a bulleted list of applied filters "Sorry, I can't find any changes files matching the following filters: " Such that the applied filters list is scannable.

I will leave it up to you if you want to work on those with out a solid yes.

Gr8z commented:
Thank you @tidy-dev, I'm happy to work on the suggested improvements while your team continues the discussion:

- Update terminology to "Included in commit/Excluded from commit"
- Add a "Clear all" button for better UX
- Implement usage metrics tracking using `incrementMetric`
- Improve the "No results" message with a scannable bulleted list of applied filters

I'll start working on these changes to have them ready for when you guys make a final decision in mid-June.

Gr8z committed:
Merge remote-tracking branch 'upstream/development' into development

Gr8z committed:
Complete comprehensive filter functionality for changes list

- Update terminology from "staged/unstaged" to "included in commit/excluded from commit"
- Add clear filters functionality with buttons in dropdown and empty state
- Implement comprehensive metrics tracking for all filter operations
- Enhance no results message with bulleted list of active filters
- Add responsive empty state design with proper container queries
- Fix hidden changes warning to work with all filter types
- Add extensive test coverage for clear filters functionality
- Verify OR logic implementation in filter operations

This completes PR #20537 with full feature implementation and test coverage.

Gr8z commented:
Hi @tidy-dev 

I have added the changes that you requested. Please take a look now.

- Update terminology from "staged/unstaged" to "included in commit/excluded from commit"
- Add clear filters functionality with buttons in dropdown and empty state
- Implement comprehensive metrics tracking for all filter operations
- Enhance no results message with bulleted list of active filters
- Add responsive empty state design with proper container queries
- Fix hidden changes warning to work with all filter types
- Add extensive test coverage for clear filters functionality

### Screenshots

![image](https://github.com/user-attachments/assets/b819361d-affb-409c-b715-80464e373bdc)
![image](https://github.com/user-attachments/assets/61d9a6fc-61a5-40cb-b10b-efd9472e6818)

tidy-dev commented on app/styles/ui/changes/_changes-list.scss:
```diff
@@ -276,3 +282,74 @@
+  flex: 1;
+  flex-direction: column;
+  align-items: center;
+  justify-content: center;
```
```suggestion
```

In our other lists, we keep the no results messaging near the top of the list. Since we are aligning to those designs, would like follow it with this aspect as well.

tidy-dev commented on app/styles/ui/changes/_changes-list.scss:
```diff
@@ -276,3 +282,74 @@
+    white-space: pre-line;
+    line-height: 1.4;
+    flex-shrink: 1; // Allow subtitle to shrink if needed
+    overflow: hidden;
```
```suggestion
```

We don't want to clip content - see comment on the `.no-changes-filtered` rule

tidy-dev commented on app/styles/ui/changes/_changes-list.scss:
```diff
@@ -276,3 +282,74 @@
+  text-align: center;
+  padding: var(--spacing);
+  box-sizing: border-box;
+  overflow: hidden; // Prevent scrollbars
```
```suggestion
  overflow-x: hidden;
  overflow-y: auto; 
```

:D We want scrollbars. Even with the nice, hide image on small screens that you added, we are still clipping the content of the no results message if there are multiple filters applied. Additionally, if a user has the screen applies the zoom feature, the content is clipped. Thus, we do want the scrollbars in this case.

Gr8z replied:
Great catch! You're right - we should allow vertical scrolling for accessibility and zoom scenarios. I'll update this.

tidy-dev commented on app/src/lib/app-state.ts:
```diff
@@ -776,6 +776,18 @@ export interface IChangesState {
 
   /** The state of the changes list filter of included changes. */
   readonly includedChangesInCommitFilter: boolean
```
This is something I considered doing in the first iteration, but decided to wait until adding more filters.

But what do you think of adding a :
`FileListFilterState` that houses all these filters as well as the `filterText`?

I doubt we will do this anytime soon, but... it is not an unreasonable jump to want these filters applied in the history view. A reusable interface would be handy for cutting down as must copy/pasting if we were to consider that in the future.

It also means that the `_setFilterX` methods in the app-store can be reduce to one method.

tidy-dev replied:
It also makes it so when we pass it down the hierarchy tree, it is just one thing to read and comprehend rather than a list of properties. (important if you are not interested in the filters when reading the prop list :D)

Gr8z replied:
Creating a `FileListFilterState` interface would definitely improve maintainability. This would reduce the number of `_setFilterX` methods in `app-store` to a single `_setFileListFilterState` method.

tidy-dev commented on app/src/lib/stats/stats-store.ts:
```diff
@@ -242,6 +242,11 @@ const DefaultDailyMeasures: IDailyMeasures = {
   typedInChangesFilterCount: 0,
   appliesIncludedInCommitFilterCount: 0,
+  appliesExcludedFromCommitFilterCount: 0,
+  appliesNewFilesFilterCount: 0,
```
Just for future proofing, Can we add the word `Changes` in from of Filter for all these files status stats since there are files lists elsewhere in the app. e.g. `appliesNewFilesChangesFilterCount`

.. The Include/Exclude from commit should be unique enough since that is something you do in the changes list.

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -340,6 +367,53 @@ export class FilterChangesList extends React.Component<
+  /** Calculate the count of included in commit files (files with DiffSelectionType.All or DiffSelectionType.Partial) */
+  private getIncludedFilesCount = memoizeOne(
+    (workingDirectory: WorkingDirectoryStatus) => {
+      return workingDirectory.files.filter(
```
Hmm.. Interesting change here. I suppose this change of not showing the count as applied to the text filtered rows is due the filter being an "OR" as apposed to an "AND". 

Meaning, now the filter would be additive in nature'.

Example:
Given the "Included in commit" and the "Excluded in commit" are both applied and you have 1 checked file.

Under "AND" filtering, the 1 checked file does not show because it is eliminated by the "Excluded from commits" filter.

Under "OR" filtering, they both show as it is included with the "Included in commit" filter.

I am still wrapping my mind around which makes most sense and if there is a reason to pick one over the other. My personal feeling is that it felt odd to apply more filters and get more results. :D Thus, it is not really a filtering effect.. it isn't filtering things out. It is matching effect.

If on the other hand, lets say you have 4 files. 2 are checked and 2 are unchecked.

With "AND" style filtering, if you checked both the "Included in commit" and "Excluded from commit" filter, then you would see no files.

Additionally, upon entering the filter options with no filters applied. You would see (2) after both those filters. Then, upon applying one of the filters, the count on the other would drop to (0).

It also would mean that any filter that has a (0) after it would reduce the results to 0 as there are 0 files that fit that filter in the current filtered list. 

Other notes on the "OR", I believe that since if you have all the filter options applied, then you have all the results of the "text filter". Thus, I think if we are going to implement the Filter as an "OR" than all the boxes should be checked on open and then you can "filter" by unchecking things you don't want in the list. 

Both directions can be logically valid. Thus, I return to what are the use cases of the filter, and which approach best handles those use cases.

Most common request reported scenarios:
1). My development software auto generates lots of files. I want to quickly find the ones that I worked on and commit changes from them. 
2). My workflow is that I am working for a long time and have modified lots of files since I last committed and I now want to chunk them out into logical commits. I want to quickly narrow down to the ones I want to commit.
3) I have selected a bunch of stuff from a long list. I want to quickly see what I am committing. 

To me these are users trying to narrow down results or filter - AKA "AND". 

Other thoughts: But, I do not believe this makes sense to do in this PR unless we got an overwhelming request for it... but, could have a filter option that is, "Apply filters with logical OR".  Because, I think this is a matter of how each individual thinks about finding things :D

Thoughts? Any compelling reason for a logical "OR" over "AND" that I am just overlooking?

tidy-dev replied:
TLDR:  I think we should use "AND"

Gr8z replied:
@tidy-dev Thank you for this thoughtful analysis of AND vs OR logic! You've raised an excellent point that I wanted to discuss further.

I completely agree that AND logic is more intuitive for the use cases you've outlined. However, there's a fundamental issue we need to address:

With pure AND logic, certain filter combinations become logically impossible:
- A file cannot be both "New" AND "Modified" (mutually exclusive states)
- A file cannot be both "Included in commit" AND "Excluded from commit" (either staged or not)
- Similarly, a file cannot be "New" AND "Deleted"

This means selecting both "New" and "Modified" with AND logic would always show 0 results, which might confuse users who expect to see both new and modified files.

AND logic would only be useful when "new, modified, deleleted" filters are combined with "included, exluded" filters. 

What are your thoughts on handling these mutually exclusive states? In my opinion, if we were to go with AND, we would need some sort of hybrid approach.

tidy-dev replied:
> This means selecting both "New" and "Modified" with AND logic would always show 0 results, which might confuse users who expect to see both new and modified files.

I think this makes sense. As a user, I am narrowing the results and a file cannot be New or Modified so yes there should be 0 in that case.

I think the user story is "I am crafting a commit. I wan't to now include the change that added some file. So I check the "New" status filter so I can filter down to just those ones".

> AND logic would only be useful when "new, modified, deleted" filters are combined with "included, excluded" filters.

I see what you are saying, as you can't say "Show me Not New ".  I see that is another potential filter vector. But, I do not think we should add them in this PR and what options you have with AND logic is a great iteration on the filter work.  _There are other potential ways of handling it like having a USE OR option if we find that users are looking for this ability through user feedback_

Gr8z replied:
Makes sense - I've updated the filter logic to use AND instead of OR. With this change, when multiple filters are selected, only files that match ALL active filters will be shown, which should provide more intuitive narrowing of search results as you suggested.

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -1456,35 +1661,93 @@ export class FilterChangesList extends React.Component<
-    return `Sorry, I can't find any changed files${filterTextMessage}${conjunction}
-        ${includedCommitText}.`
+    const bulletedList = activeFilters.map(filter => `• ${filter}`).join('\n')
+    return `Sorry, I can't find any changed files matching the following filters:\n${bulletedList}`
```
![Image](https://github.com/user-attachments/assets/0297194c-d969-497e-8883-1a45b3ee7a25)

I know I recommended the bullet points (I wasn't thinking about it being centered). But.. I think we need another approach for our more visual users as centered aligned bullets feel a bit rough. 

I am torn on a recommendation tho, I believe the bullet list does help make clear what is impacting the no results message, but likely unless we have a visually "not part of a sentence" UI here, then they need to be comma delimited.

Gr8z replied:
I'll switch from bullet points to a comma-delimited list, which will look cleaner while maintaining readability. You can then compare if it looks better than bulleted centered list.

tidy-dev reviewed (changes requested):
Thank you for all your work thus far. 💖 This is generally looking good.

Lots of feedback for you for my initial review on the logical "OR" approach. Feel free to ask questions if anything isn't clear.

_FYI, going to gather some feedback from a coworker on the unit tests as they just did a bunch of work in that realm_

Gr8z committed:
Refactor filter state to use FileListFilterState interface

Gr8z committed:
Update CSS for better overflow handling and top alignment

Gr8z committed:
Add Changes prefix to filter stats naming

Gr8z committed:
Change no results message to comma-delimited format

Gr8z committed:
Update tests for FileListFilterState interface

Gr8z committed:
fix unit tests and eslint

Gr8z committed:
Clean up CSS

Gr8z commented:
- Updated filter state interface - Refactored to use the new `FileListFilterState` interface for better type safety
  - Improved filter statistics naming - Added "Changes" prefix to filter stats for clarity
  - Enhanced no results message - Changed to comma-delimited format for better readability when multiple filters are active
  - Improved CSS overflow handling - Updated styles for better overflow behavior and top alignment in the filter dropdown

![image](https://github.com/user-attachments/assets/951ec57b-5c66-42cd-a729-b5cbd3108c0c)
![image](https://github.com/user-attachments/assets/9b8a67a4-ec19-47fb-9d32-54f558ad78f7)


I am waiting for a decision on the [AND and OR discussion](https://github.com/desktop/desktop/pull/20537#discussion_r2151791898) for me to add those changes (or not).

niik commented on app/test/unit/app-store-filter-test.ts:
```diff
@@ -0,0 +1,608 @@
+const allSelected = DiffSelection.fromInitialSelection(DiffSelectionType.All)
+const noneSelected = DiffSelection.fromInitialSelection(DiffSelectionType.None)
+
+describe('App Store Filter Functionality', () => {
```
It seems that all the tests (I only skimmed it) in this file operate on the same fundamental premise of verifying that `updateChangesState` does in fact update. Since `updateChangesState` is straightforward in nature, only applying changes with no specific logic related to the properties here I don't think these tests add any value and we should strip them from this PR before merging.

If there's tests in here that I've missed that do more than setting state properties and then retrieving them to verify that they've been set we can keep them and review them separately

niik commented on app/test/unit/changes-filter-integration-test.ts:
```diff
@@ -0,0 +1,1103 @@
+  let testFiles: WorkingDirectoryFileChange[]
+  let workingDirectory: WorkingDirectoryStatus
+
+  beforeEach(() => {
```
I've only skimmed this file as well but it looks to me like the only way any of these tests would fail is if we modify the tests themselves given that the filter logic is mocked and we don't rely on any other application logic (I mean it would break if we broke something like WorkingDirectoryFileChange but that's not what these tests are meant to verify). I propose we exclude these tests from the PR.

As an aside, performance tests like the one included here is something we try to avoid. The test as written now relies on a change being made to the filtering logic (which we've already established is mocked) such that the filtering becomes expensive enough that one of us developer would question the test suite run time and go looking.

niik commented on app/test/unit/clear-filters-metrics-test.ts:
```diff
@@ -0,0 +1,340 @@
+import { describe, it } from 'node:test'
+import assert from 'node:assert'
+
+describe('Clear Filters Metrics Tests', () => {
```
Similar to the other two files these functions only verify that Node's Mock method work, not touching any application logic so by following the same logic as in the previous the only way we'd break these tests is if Node's mock method would fail or if we changed the tests themselves. Let's not take this.

niik commented on app/test/unit/dispatcher-filter-test.ts:
```diff
@@ -0,0 +1,320 @@
+      const filterText = 'README'
+      const filterBoolean = true
+
+      assert.strictEqual(typeof filterText, 'string')
```
I don't see any point in any of the tests in this file.

niik commented on app/test/unit/filter-changes-list-count-methods-test.ts:
```diff
@@ -0,0 +1,417 @@
+  .withLineSelection(1, false)
+  .withLineSelection(3, false)
+
+describe('FilterChangesList Count Methods Tests', () => {
```
Same as the others, no application logic actually being tested as far as I can see, let's remove it.

niik commented on app/test/unit/filter-changes-list-ui-interactions-test.ts:
```diff
@@ -0,0 +1,808 @@
+const allSelected = DiffSelection.fromInitialSelection(DiffSelectionType.All)
+const noneSelected = DiffSelection.fromInitialSelection(DiffSelectionType.None)
+
+describe('FilterChangesList UI Interactions Tests', () => {
```
Same as the others, no application logic actually being tested as far as I can see, let's remove it.

niik reviewed (dismissed):
I was asked to review the tests in this PR and the sheer volume of tests here made it impossible for me to go through them all but I've skimmed through a lot of them and I see a lot of the same patterns so I've reviewed the patterns and not each test so I apologize if I've missed something of substance but the tests that I've looked at are all self-contained in so far that the only way they could break is if we changed the tests themselves.

Gr8z committed:
Remove test files that don't test actual application logic

Gr8z committed:
Remove changes-filter-test.ts - only tests setter/getter functionality

This test file follows the same problematic pattern as the others removed
earlier. It only verifies that state values can be set and retrieved,
without testing any actual filtering logic or application behavior.

Gr8z reviewed (commented):
Hi @niik,

I've removed all the test files as you suggested - they were indeed just testing mock implementations rather than actual application logic.

The actual filtering logic in this PR is implemented as private methods within the React component:
- `applyFilters` method - contains the core filtering logic
- Various count methods (getNewFilesCount, getModifiedFilesCount, etc.) - private memoized methods

Since these are private methods in a React component, they cannot be directly imported and tested. How should we approach testing this functionality?

Gr8z committed:
Change filter logic from OR to AND for narrowing search results

- Modified applyFilters method to use AND logic instead of OR logic
- Files must now satisfy ALL active filters to be displayed
- This ensures filtering narrows down results rather than expanding them
- Addresses reviewer feedback to make filtering more intuitive for users

When multiple filters are selected, the list will show only files that match all selected criteria, providing more precise filtering.

tidy-dev commented:
>The actual filtering logic in this PR is implemented as private methods within the React component:
> applyFilters method - contains the core filtering logic
> Various count methods (getNewFilesCount, getModifiedFilesCount, etc.) - private memoized methods
> Since these are private methods in a React component, they cannot be directly imported and tested. How should we approach testing this functionality? 

When considering testing, we look to test functions with unique complex logic, and in doing so, will isolate that logic into a function that is exportable. For example, if you wanted to test "apply filters", you would have to rewrite it such that all the "props" properties that are used in it our taken as parameters instead. Then, the method could live outside the react component (like at the top of the file under the file imports or in it's own .ts file). 

N.B. GitHub Desktop takes a pragmatic approach to testing and we do not test to get 100% coverage, but to test the bits of complex logic in a targeted concise manner. We do not look to have dozens of tests for something like this.  In fact, the count methods are straight forward enough, that I do not think the overhead of making them tested is worth it. But, the `applyFilters` as it is complex and has many logical turns it and likely could change if we added new filters down the road, makes a likely candidate to be tested.

Gr8z committed:
added new tests and extracted functions

Gr8z committed:
adjust filters fix

tidy-dev commented on app/src/lib/stats/stats-database.ts:
```diff
@@ -607,6 +607,21 @@ export interface IDailyMeasures {
+  appliesDeletedFilesChangesFilterCount: 0
+
+  /** The number of times the user clears all filters */
+  appliesClearAllFiltersCount: 0
```
```suggestion
  appliesClearAllChangesListFilterCount: 0
```

^ Apply this as needed to make this stat unique.

tidy-dev commented on app/src/lib/app-state.ts:
```diff
@@ -771,11 +771,31 @@ export interface IChangesState {
+  readonly filterDeletedFiles: boolean
+
+  /** Whether to filter and show only excluded from commit files */
+  readonly filterExcludedFiles: boolean
```
Nit: "filterExcludedFiles" is the opposite of "includedChangesInCommitFilter" but they feel very different. Be nice to use consistent filter names across all filters. Additionally, "Exclude" feels to generic? 

Another Nit: We are grouping these in a `IFileListFilterState`, I do not think they need the word "filter" in them. Thus, we can be more concise with names like "isIncludeInCommit", "isExcludedFromCommit", "isNewFile", "isModifiedFile", "isDeletedFile".  (The "is" helps denote they are boolean)

tidy-dev replied:
See https://github.com/desktop/desktop/pull/20537/files#r2161555672, reuse the `IFileListState`here (or vice versa) We only need one interface. :)

tidy-dev commented on app/src/lib/stores/app-store.ts:
```diff
@@ -8313,21 +8314,54 @@ export class AppStore extends TypedBaseStore<IAppState> {
     this.emitUpdate()
   }
 
+  // Legacy methods for backwards compatibility
+  public _setChangesListFilterText(repository: Repository, filterText: string) {
+    this._updateFileListFilter(repository, { filterText })
+  }
+
```
```suggestion
```

No need to keep around unused methods.

tidy-dev replied:
(Basing this off the comment, if it is used, then remove only the comment)

Gr8z replied:
We are still using this function, lets remove this comment.

tidy-dev commented on app/test/unit/filter-changes-logic-test.ts:
```diff
@@ -0,0 +1,478 @@
+    })
+  })
+
+  describe('getCheckBoxValueFromIncludeAll', () => {
+    it('should return On for true', () => {
+      assert.equal(getCheckBoxValueFromIncludeAll(true), CheckboxValue.On)
+    })
+
+    it('should return Off for false', () => {
+      assert.equal(getCheckBoxValueFromIncludeAll(false), CheckboxValue.Off)
+    })
+
+    it('should return Mixed for null', () => {
+      assert.equal(getCheckBoxValueFromIncludeAll(null), CheckboxValue.Mixed)
+    })
+  })
```
Wouldn't bother testing or extracting `getCheckBoxValueFromIncludeAll` for testing. Method is simple and not likely to change.

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -94,18 +109,6 @@ const StashIcon: OcticonSymbolVariant = {
 
 const GitIgnoreFileName = '.gitignore'
 
-function getCheckBoxValueFromIncludeAll(includeAll: boolean | null) {
```
Mentioned elsewhere, I wouldn't bother extracting this for the sake of testing. Only worth extracting if we were to reuse it elsewhere.

tidy-dev commented on app/src/ui/changes/sidebar.tsx:
```diff
@@ -459,9 +459,19 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
           aheadBehind={this.props.aheadBehind}
           accounts={this.props.accounts}
-          filterText={this.props.changes.filterText}
+          filterText={this.props.changes.fileListFilter.filterText}
```
We can just pass the "filterListFilter" object instead of splitting it all out here.

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -226,6 +229,18 @@ interface IFilterChangesListProps {
   readonly includedChangesInCommitFilter: boolean
 
+  /** Whether to filter and show only new files */
+  readonly filterNewFiles: boolean
```
See other comment - No need for these to be split out, only need to pass down the filter state object.

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -268,30 +283,27 @@ export class FilterChangesList extends React.Component<
+  private filterListRef =
+    React.createRef<AugmentedSectionFilterList<IChangesListItem>>()
 
   private isCommittingFileHiddenByFilter = memoizeOne(
```
Should just extract the whole thing as a memoized method into the `filter-changes-logic.ts` file. The tests should be distinct enough that the memoization won't impact them.

tidy-dev replied:
Still needs completely extracted with the `meoizeOne`

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -300,51 +312,24 @@ export class FilterChangesList extends React.Component<
-          f => f.status.kind === AppFileStatusKind.Untracked
-        )
+    ) => 
+      getCheckAllValue(workingDirectory, rebaseConflictState, filteredItems)
```
Not terribly opposed to this being tested, but your PR is already touching a lot of surface area. Lets not separate it out this just for the sake of adding tests to it.

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,315 @@
+/**
+ * Calculate the count of new files (including untracked)
+ */
+export function getNewFilesCount(workingDirectory: WorkingDirectoryStatus): number {
```
As I stated previously, I do not believe the overhead of separating these for the sake of testing is worth the maintenance overhead for such simple methods. Please keep these in the fitler-changes-list.tsx and remove the unit tests for all the count methods.

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,315 @@
+    return undefined
+  }
+
+  const filterList = activeFilters.join(', ')
```
Since this is now sentence form, we should have an ", and" before the last one in the list (when more than one)

tidy-dev commented on app/test/unit/filter-changes-logic-test.ts:
```diff
@@ -0,0 +1,478 @@
+    })
+  })
+
+  describe('hasActiveFilters', () => {
```
Tests not worth the overhead. Please remove test for `hasActiveFilters`

tidy-dev commented on app/test/unit/filter-changes-logic-test.ts:
```diff
@@ -0,0 +1,478 @@
+    })
+  })
+
+  describe('countActiveFilters', () => {
```
Again - too simple to be worth bothering to have tests.

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,315 @@
+  filters: IFileFilterState
+): boolean {
+  return (
+    filterText !== '' ||
+    filters.includedChangesInCommitFilter ||
+    filters.filterNewFiles ||
+    filters.filterModifiedFiles ||
+    filters.filterDeletedFiles ||
+    filters.filterExcludedFiles
```
```suggestion
    filterText !== '' || countActiveFilters(filters) > 0
```

Could simplify/reuse

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,315 @@
+/**
+ * Count the number of active filters
+ */
+export function countActiveFilters(filters: IFileFilterState): number {
```
Maybe this should be `countActiveFilterOptions` to specify that it doesn't check the `filterText` filter. The method documentation should mention that  as well.

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,315 @@
+  filters: IFileFilterState
+): boolean {
+  // Check if any filters are active
+  const hasAnyActiveFilter =
```
List should use the utility method for checking `hasAnyActiveFilters`

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,315 @@
+ * Apply filters to determine if a file should be shown
+ * Uses AND logic - file must satisfy ALL active filters
+ */
+export function applyFilters(
```
Nit: Maybe this should be `applyFilterOptions` and call out in the documentation above that this is applied after the `filterText` has been applied.  To make it clear that `filterText` and the filters in the filter options are handled separately

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -1454,37 +1533,53 @@ export class FilterChangesList extends React.Component<
-
-    return `Sorry, I can't find any changed files${filterTextMessage}${conjunction}
-        ${includedCommitText}.`
+    return getNoResultsMessage(this.props.filterText, {
```
Nit: Once the filter state object is passed down as a prop, I think it won't make sense to separate out the `filterText` and simply pass in a filters object.

tidy-dev reviewed (changes requested):
I have left a hand full of comments and they are interacting with each other so I am going to pause on reviewing for now. Happy to take another look after another iteration. 

This is looking good. Just a little more on the refactoring/unit test sides of things than needed.

Gr8z committed:
code review fixes

Gr8z commented:
@tidy-dev, thanks for the extensive review once again. I have addressed them all with the best of my ability. Let me know if I missed anything.

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -263,35 +260,45 @@ function getSelectedItemsFromProps(
 }
 
+/** Helper to convert IFileListFilterState to IFileFilterState */
+function mapToFileFilterState(
```
Instead of having to map the filter objects, could you update so that the app state and the file list both use the `IFileFilterState` interface?

(I believe you would just need to add the `filterText` property)

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,179 @@
+}
+
+/** Interface for a changes list item */
+export interface IChangesListItem {
```
No need for a duplicate `IChangesListItem` interface. Please remove and import from the filter-changes-list.tsx.

tidy-dev commented on app/src/ui/changes/filter-changes-logic.ts:
```diff
@@ -0,0 +1,179 @@
+/**
+ * Check if any files being committed are hidden by the current filter
+ */
+export function isCommittingFileHiddenByFilter(
```
This should using `memoizeOne` as it used to in the`filter-changes-list.tsx`

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -380,6 +433,21 @@ export class FilterChangesList extends React.Component<
   }
 
+  /** Get checkbox value from includeAll status */
+  private getCheckBoxValueFromIncludeAll(
```
No need to move this inside the component. We can just keep the old method referenced in https://github.com/desktop/desktop/pull/20537/files#r2154607813

Gr8z replied:
I am not sure I fully understand this. In a previous review, you mentioned that we don't need to remove it from the component because it's not worth it.

tidy-dev replied:
Sorry for my lack of clarity. Keep it as it was before this PR and do not test it.

Gr8z replied:
noted 👍🏼

tidy-dev commented on app/src/ui/changes/filter-changes-list.tsx:
```diff
@@ -1454,53 +1592,132 @@ export class FilterChangesList extends React.Component<
-
-    return `Sorry, I can't find any changed files${filterTextMessage}${conjunction}
-        ${includedCommitText}.`
+    return getNoResultsMessage(
```
Can we call the imported `getNoResultsMessage` directly whenever we need it?

tidy-dev reviewed (changes requested):
Thank you for those changes. Just a couple more nits I think. Will do some functional testing today.

Gr8z committed:
Merge remote-tracking branch 'upstream/development' into development

Gr8z committed:
Refactor file filter interfaces and add memoization

Gr8z committed:
Merge remote-tracking branch 'upstream/development' into development

Gr8z commented:
- Removed duplicate IFileFilterState interface from filter-changes-logic.ts
- Now using single IFileListFilterState interface from app-state.ts throughout the codebase
- Exported IChangesListItem from filter-changes-list.tsx instead of duplicating it
- Kept getCheckBoxValueFromIncludeAll outside the component as requested
- Direct function calls to getNoResultsMessage without wrapper

Gr8z commented:
@tidy-dev do you have any more updates to move this forward?

tidy-dev committed:
Create Filter Options Component

- Make counts reflect "AND" logic
- Simply some methods that can just consume the filter state

tidy-dev committed:
Refactor filter logic to use FileChange methods

Replaces direct property checks and enum comparisons with method calls on the FileChange object for filtering changes. This improves encapsulation and readability by delegating status checks to the change object's methods.

tidy-dev reviewed (changes requested):
@Gr8z Sorry it took me a while to get back to you!

There are few things we haven't addressed from some of my previous feedback. I apologize for not summarizing it in one place.  I decided to see if [I could wrap that up real quick for you so we can get this merged and found a few places we can make the code a little cleaner/more efficient](https://github.com/desktop/desktop/commit/1171f3a4001c4c99c93f40169b44a891b87265d1) Take a look and see if you agree with the changes and pull them in if so. :) 

Things it does previously discussed or inferred:
- Makes the filter options count reflect the number of changes that are currently in the list that match the filter. (respecting the AND logic) (So that when you apply a filter the list will narrow to the number shown)
- Makes a couple of the methods not have a redundant `filterText` parameter since that is now on the filter state object.
- Removes the redundant get fileFilterState method in favor of directly referring to the props since there is no mutation of the props state happening.

In addition, 
- Consolidated the count methods into a single memoized method such that we don't need to loop through all the filtered values for every count metric as they should all be recounted anytime the filtered values or working directory changes. 
- I moved the filter popover logic into it's own component to group it's logic together. (Not necessary, just seems easier to digest to me and possibly something I should have done originally when introducing the popover)
- Added some helper methods on to the FileChange classes to make the filter logic a little cleaner.

Gr8z commented:
Nice @tidy-dev, took a look at [your changes](https://github.com/desktop/desktop/commit/1171f3a4001c4c99c93f40169b44a891b87265d1) and it looks super clean. I apologize if I missed any of your comments or misinterpreted them.  

One thing, though, I am not sure if showing the count this way is the right way to go. The filter option counts should show the independent counts, not the combined AND counts.

![image](https://github.com/user-attachments/assets/0765893e-11f7-48de-9692-8434927e9f5b)

I have two files here to be committed, but you can't tell because the count is 0.

tidy-dev commented:
> The filter option counts should show the independent counts, not the combined AND counts.

I can see where you are coming from. And, I view them as filters and I think the number should help the user use the filter. Just like on any e-commerce site, if you put a search term in, and you receive 100 results. There are additional filters usually provided such as color or rating. Those filters often tell you of the 100 results present 60 are blue and 50 have 4 star rating. But.. if I apply the blue filter, now my list has 60 items and now the ratings counts are updated to say there are 10 items with 4 star ratings in the list (that are blue and match the search term). 

If we flip this and leave them as independent counts, say I have 100 files in the list and 50 are new, 30 are modified, and 20 are deleted.  50 of them are checked and 50 are not. I open the filter and see those counts. I apply the "New files" filter. The lists filters down to the 50 files. I open the filter options again and I still see 50 are checked and 50 are not checked.. but there are only 50 files in the list.. so I have no idea how many are going to show if I checked "Included in commit". 

Where as if the filter options inform you of how many files the filter will filter to, when I opened it the second time it should look like "New files" (50), modified/deleted (0), and could look something like  "Included in commit" (10) and "Excluded from commit" (40). Now I know if I apply the modified or deleted filter, there will be 0 left in the list and if I apply the "Included in commit" there will be 10 left in the list. 

Additionally, that is the way the filter works today, if you have 100 files in the list and apply a "test" string then open the filter options the "Included in commit" count will be something < 100 even if all 100 were checked prior to the search string.

I propose for now we keep the logic inline with the first iteration and see what kind of community feedback we get for it. Another iteration could be completely switching this logic or could be something like having a (0/100) where the first number shows the filter applied and the second number is the independent count. We do this for the string filter.

Gr8z commented:
The e-commerce analogy misses a key difference in user intent. In e-commerce, users refine searches to shrink choices and discover products from an unknown catalog. In Git clients, users aren't discovering files - they're verifying and organizing work they've already done. They need transparency about repository state to maintain accurate mental models, not guidance on what subset to focus on.

When I see "Included in commit (0)" but know I have staged files, it's unclear whether the filter is working correctly, there's a bug, or I've misunderstood the interface. Independent counts would show "Included in commit (2)" regardless of other filters, making it clear that 2 files are staged and the current 0 result is due to filter intersection. The dual format "Included in commit (0/2)" could provide both contextual and absolute info.

is there a beta testing branch with an enough user pool to test both approaches and gather actual feedback before committing to either pattern?

tidy-dev commented:
> e-commerce, users refine searches to shrink choices

That is my analogy. Using the filter to shrink (filter) choices to smaller subset to commit.. or just find a file to look at in the diff.  This component is a tool to accomplish filtering not about summarizing what they already know. Tho, I do agree that it is not perfect analogy as you note about unknown total list versus a finite known list. I revert back to the purpose of the list. The purpose is to allow users to either (1) select a file to review the diff or (2) check files to include in the commit. The filter was made to assist in those purposes when the list is long enough to be hard to scan or you must scroll to find a file. Thus, it purpose is to narrow the list.. not provide ancillary information about the list. 

> The dual format "Included in commit (0/2)" could provide both contextual and absolute info.

Again to your point, I can see someone using the ancillary information as form of confirmation that the list reflects what they expect it to before they commit so it can also be useful in assisting in commit files.

The reason I am hesitant in implementing this solution now is a prior design discussion my team held where we considered it on the first iteration, but chose to limit the design to necessary information for the filter function instead of more numbers to figure out what they represent as the space is too limited to really communicate what they represent.


>  beta testing branch

Great idea! Check out the `feature-flag.ts` and add a new one to encapsulate these new filters and we will hold to the [beta channel](https://desktop.github.com/beta/) for a few months to have a chance to gather some feedback. The beta channel is not a large cohort so it takes time to reasonably obtain some feedback.

Gr8z commented:
I completely understand your perspective, and I trust your judgment. The space constraints you mention are also valid concerns.

Either we can implement a feature flag and wait, or, since, in my opinion, this isn't a disruptive feature, users can choose not to use it or completely hide it from view. We can then merge it as is and get fast feedback.

tidy-dev commented:
> Not disruptive.

I agree. If you like to just pull the changes in from my commit, I will take another look and maybe we can get this merged. :)

Gr8z committed:
Merge remote-tracking branch 'upstream/development' into development

Gr8z commented:
Sounds good, Done!

tidy-dev commented on app/test/unit/filter-changes-logic-test.ts:
```diff
@@ -0,0 +1,238 @@
+// Helper function to create a test file
+function createTestFile(
+  path: string,
+  status: AppFileStatusKind,
```
```suggestion
  kind:
    | AppFileStatusKind.New
    | AppFileStatusKind.Modified
    | AppFileStatusKind.Deleted,
```

tidy-dev commented on app/test/unit/filter-changes-logic-test.ts:
```diff
@@ -0,0 +1,238 @@
+    path,
+    status: { kind: status },
+    selection,
+  } as WorkingDirectoryFileChange
```
```suggestion
  return new WorkingDirectoryFileChange(path, { kind }, selection)
```

tidy-dev commented on app/test/unit/filter-changes-logic-test.ts:
```diff
@@ -0,0 +1,238 @@
+// Helper function to create a test changes list item
+function createTestItem(
+  path: string,
+  status: AppFileStatusKind,
```
```suggestion
  status:
    | AppFileStatusKind.New
    | AppFileStatusKind.Modified
    | AppFileStatusKind.Deleted,
```

tidy-dev reviewed (changes requested):
Working as expected. ✨ Super close to merge, but build is failing due to test relying on a object cast instead of instantiating the class as well as a linter error. I have added suggestions to fix the unit test error and then should just need to run `yarn lint:fix` to address the formatting issue.

Gr8z committed:
Update filter-changes-logic-test.ts

Gr8z committed:
yarn lint

Gr8z commented:
Updated the tests, they are all passing now @tidy-dev

tidy-dev reviewed (approved):
🎉  Thanks for all the iterations and thoughtful discussion @Gr8z!

materesarmt-coder reviewed (commented):
#
